sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RetryPolicy.java

```diff
@@ -59,58 +77,103 @@ public RetryPolicy() {
-     * delay provided in response header.
-     * @param retryAfterTimeUnit The time unit to use when applying the retry delay. null is valid if, and only if,
-     * {@code retryAfterHeader} is null.
-     *
-     * @throws NullPointerException If {@code retryStrategy} is null or when {@code retryAfterTimeUnit} is null and
-     * {@code retryAfterHeader} is not null.
-     */
-    public RetryPolicy(RetryStrategy retryStrategy, String retryAfterHeader, ChronoUnit retryAfterTimeUnit) {
-        this.retryStrategy = Objects.requireNonNull(retryStrategy, "'retryStrategy' cannot be null.");
         this.retryAfterHeader = HeaderName.fromString(retryAfterHeader);
         this.retryAfterTimeUnit = retryAfterTimeUnit;
         if (!isNullOrEmpty(retryAfterHeader)) {
             Objects.requireNonNull(retryAfterTimeUnit, "'retryAfterTimeUnit' cannot be null.");
         }
     }
 
-    /**
-     * Creates a {@link RetryPolicy} with the provided {@link RetryStrategy}.
-     *
-     * @param retryStrategy The {@link RetryStrategy} used for retries.
-     *
-     * @throws NullPointerException If {@code retryStrategy} is null.
+    public RetryPolicy() {
+        this(null, null);
+    }
+
+    @Override
+    public HttpResponse process(HttpRequest httpRequest, HttpPipelineNextPolicy next) {
+        return attempt(httpRequest, next, 0, null);
+    }
+
+    /*
+     * Determines the delay duration that should be waited before retrying using the well-known retry headers.
      */
-    public RetryPolicy(RetryStrategy retryStrategy) {
-        this(retryStrategy, null, null);
+    public static Duration getWellKnownRetryDelay(Headers responseHeaders, int tryCount, RetryStrategy retryStrategy,
+                                                  Supplier<OffsetDateTime> nowSupplier) {
+        Duration retryDelay = ImplUtils.getRetryAfterFromHeaders(responseHeaders, nowSupplier);
+        if (retryDelay != null) {
+            return retryDelay;
+        }
+
+        // None of the well-known headers have been found, return the default delay duration.
+        return retryStrategy.calculateRetryDelay(tryCount);
+    }
+
+    public Integer getMaxRetries() {
+        if (maxRetries == null) {
+            return DEFAULT_MAX_RETRIES;
+        }
+        return maxRetries;
+    }
+
+    public RetryPolicy setMaxRetries(Integer maxRetries) {
+        if (maxRetries < 0) {
+            throw LOGGER.logThrowableAsError(new IllegalArgumentException("Max retries cannot be less than 0."));
+        }
+        this.maxRetries = maxRetries;
+        return this;
+    }
+
+    public RetryStrategy getRetryStrategy() {
+        if (retryStrategy == null) {
+            retryStrategy = DEFAULT_RETRY_STRATEGY;
+        }
+        return retryStrategy;
+    }
+
+    public RetryPolicy setRetryStrategy(RetryStrategy retryStrategy) {
+        this.retryStrategy = Objects.requireNonNull(retryStrategy, "'retryStrategy' cannot be null.");
+        return this;
```

Comment:
Shouldn't the RetryPolicy be immutable? It'd be really strange behavior for the retrying logic to change mid request or to have the max retry count change after more retries than allowed have been attempted.

Reply:
The only way to set a retry policy is via the builder and don't see how users can have access to the retry policy object to alter anything mid request.